File: lib/C110PSerial/C110PSerial.cpp
#include "C110PSerial.h"

bool C110PSerial::send(const C110PCommand& msg)
{
    uint8_t buffer[MAX_SIZE] = {0};
    pb_ostream_t stream = pb_ostream_from_buffer(buffer, sizeof(buffer));
    if (!pb_encode(&stream, C110PCommand_fields, &msg))
    {
        std::cout << "Failed to encode C110PCommand message: " << PB_GET_ERROR(&stream) << std::endl;
        return false;
    }
    m_sentMessageBuffer.add(msg);
    m_messageInfoMap[msg.id] = {this->getSafeTimestamp(), 0};
    
    size_t len = stream.bytes_written;
    uint8_t crc = crc8.calculate(buffer, len);
    std::cout << "Sending data: [";
    for (size_t i = 0; i < len; ++i) {
        std::cout << std::hex << std::uppercase << static_cast<int>(buffer[i]);
        if (i < len - 1) std::cout << " ";
    }
    std::cout << "] LEN: " << len << " CRC: " << std::hex << std::uppercase << static_cast<int>(crc) << std::dec << std::endl;
    if (m_stream->write(START_BYTE) == 0 ||
        m_stream->write(len) == 0 ||
        m_stream->write(buffer, len) == 0 ||
        m_stream->write(crc) == 0)
    {
        return false;
    }
    return true;
}

File: python/lib/C110PSerial/C110PSerial.py
from .ProtoFrame import ProtoFrame
from typing import Callable, Optional
from .proto_encode import encode_command
from .Logger import logger

C110PRegion_REGION_UNSPECIFIED = 0
C110PRegion_REGION_BODY = 1
C110PRegion_REGION_LEG = 2
C110PRegion_REGION_NECK = 3
C110PRegion_REGION_DOME = 4

C110PActuator_UNSPECIFIED = 0
C110PActuator_BODY_NECK = 1

class C110PSerial(ProtoFrame):
    def __init__(self, stream, identifier=C110PRegion_REGION_UNSPECIFIED, timeout=1000):
        super().__init__(stream, identifier, timeout)

    def send(self, msg) -> bool:
        err, buffer = encode_command(msg)
        if len(buffer) > self.MAX_SIZE:
            logger.error(f"Message too large: {len(buffer)}")
            return False
        if err:
            logger.error(f"Error encoding message: {err}")
            return False

        crc = self.crc8.calculate(buffer)
        logger.debug("Sending data: [{}] LEN: {} CRC: {:02X}".format(
            ' '.join('{:02X}'.format(b) for b in buffer),
            len(buffer),
            crc
        ))

        if not any((self.m_stream.write(bytes([self.START_BYTE])),
            self.m_stream.write(bytes([len(buffer)])),
            self.m_stream.write(buffer),
            self.m_stream.write(bytes([crc])))
        ):
            logger.error("Write failed")
            return False
        
        self.trackMessage(msg)
        return True

    def processQueue(self):
        self.readFrame()
        self.retryMessages()

    def createLedCommand(self, target, start, end, duration=0):
        cmd = {
            "id": self.getSafeTimestamp(),
            "source": self.m_regionId,
            "target": target,
            "led": {
                "start": start,
                "end": end,
                "duration": duration
            }
        }
        return cmd

    def createSoundCommand(self, target, soundId, play=False, syncToLeds=False):
        cmd = {
            "id": self.getSafeTimestamp(),
            "source": self.m_regionId,
            "target": target,
            "sound": {
                "id": soundId,
                "play": play,
                "syncToLeds": syncToLeds
            }
        }
        return cmd

    def createMoveCommand(self, target, move_target, x=0, y=0, z=0):
        cmd = {
            "id": self.getSafeTimestamp(),
            "source": self.m_regionId,
            "target": target,
            "move": {
                "target": move_target,
                "x": x,
                "y": y,
                "z": z
            }
        }
        return cmd
